\documentclass[11pt]{article}

    \usepackage[breakable]{tcolorbox}
    \usepackage{parskip} % Stop auto-indenting (to mimic markdown behaviour)
    
    \usepackage{iftex}
    \ifPDFTeX
    	\usepackage[T1]{fontenc}
    	\usepackage{mathpazo}
    \else
    	\usepackage{fontspec}
    \fi

    % Basic figure setup, for now with no caption control since it's done
    % automatically by Pandoc (which extracts ![](path) syntax from Markdown).
    \usepackage{graphicx}
    % Maintain compatibility with old templates. Remove in nbconvert 6.0
    \let\Oldincludegraphics\includegraphics
    % Ensure that by default, figures have no caption (until we provide a
    % proper Figure object with a Caption API and a way to capture that
    % in the conversion process - todo).
    \usepackage{caption}
    \DeclareCaptionFormat{nocaption}{}
    \captionsetup{format=nocaption,aboveskip=0pt,belowskip=0pt}

    \usepackage{float}
    \floatplacement{figure}{H} % forces figures to be placed at the correct location
    \usepackage{xcolor} % Allow colors to be defined
    \usepackage{enumerate} % Needed for markdown enumerations to work
    \usepackage{geometry} % Used to adjust the document margins
    \usepackage{amsmath} % Equations
    \usepackage{amssymb} % Equations
    \usepackage{textcomp} % defines textquotesingle
    % Hack from http://tex.stackexchange.com/a/47451/13684:
    \AtBeginDocument{%
        \def\PYZsq{\textquotesingle}% Upright quotes in Pygmentized code
    }
    \usepackage{upquote} % Upright quotes for verbatim code
    \usepackage{eurosym} % defines \euro
    \usepackage[mathletters]{ucs} % Extended unicode (utf-8) support
    \usepackage{fancyvrb} % verbatim replacement that allows latex
    \usepackage{grffile} % extends the file name processing of package graphics 
                         % to support a larger range
    \makeatletter % fix for old versions of grffile with XeLaTeX
    \@ifpackagelater{grffile}{2019/11/01}
    {
      % Do nothing on new versions
    }
    {
      \def\Gread@@xetex#1{%
        \IfFileExists{"\Gin@base".bb}%
        {\Gread@eps{\Gin@base.bb}}%
        {\Gread@@xetex@aux#1}%
      }
    }
    \makeatother
    \usepackage[Export]{adjustbox} % Used to constrain images to a maximum size
    \adjustboxset{max size={0.9\linewidth}{0.9\paperheight}}

    % The hyperref package gives us a pdf with properly built
    % internal navigation ('pdf bookmarks' for the table of contents,
    % internal cross-reference links, web links for URLs, etc.)
    \usepackage{hyperref}
    % The default LaTeX title has an obnoxious amount of whitespace. By default,
    % titling removes some of it. It also provides customization options.
    \usepackage{titling}
    \usepackage{longtable} % longtable support required by pandoc >1.10
    \usepackage{booktabs}  % table support for pandoc > 1.12.2
    \usepackage[inline]{enumitem} % IRkernel/repr support (it uses the enumerate* environment)
    \usepackage[normalem]{ulem} % ulem is needed to support strikethroughs (\sout)
                                % normalem makes italics be italics, not underlines
    \usepackage{mathrsfs}
    

    
    % Colors for the hyperref package
    \definecolor{urlcolor}{rgb}{0,.145,.698}
    \definecolor{linkcolor}{rgb}{.71,0.21,0.01}
    \definecolor{citecolor}{rgb}{.12,.54,.11}

    % ANSI colors
    \definecolor{ansi-black}{HTML}{3E424D}
    \definecolor{ansi-black-intense}{HTML}{282C36}
    \definecolor{ansi-red}{HTML}{E75C58}
    \definecolor{ansi-red-intense}{HTML}{B22B31}
    \definecolor{ansi-green}{HTML}{00A250}
    \definecolor{ansi-green-intense}{HTML}{007427}
    \definecolor{ansi-yellow}{HTML}{DDB62B}
    \definecolor{ansi-yellow-intense}{HTML}{B27D12}
    \definecolor{ansi-blue}{HTML}{208FFB}
    \definecolor{ansi-blue-intense}{HTML}{0065CA}
    \definecolor{ansi-magenta}{HTML}{D160C4}
    \definecolor{ansi-magenta-intense}{HTML}{A03196}
    \definecolor{ansi-cyan}{HTML}{60C6C8}
    \definecolor{ansi-cyan-intense}{HTML}{258F8F}
    \definecolor{ansi-white}{HTML}{C5C1B4}
    \definecolor{ansi-white-intense}{HTML}{A1A6B2}
    \definecolor{ansi-default-inverse-fg}{HTML}{FFFFFF}
    \definecolor{ansi-default-inverse-bg}{HTML}{000000}

    % common color for the border for error outputs.
    \definecolor{outerrorbackground}{HTML}{FFDFDF}

    % commands and environments needed by pandoc snippets
    % extracted from the output of `pandoc -s`
    \providecommand{\tightlist}{%
      \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
    \DefineVerbatimEnvironment{Highlighting}{Verbatim}{commandchars=\\\{\}}
    % Add ',fontsize=\small' for more characters per line
    \newenvironment{Shaded}{}{}
    \newcommand{\KeywordTok}[1]{\textcolor[rgb]{0.00,0.44,0.13}{\textbf{{#1}}}}
    \newcommand{\DataTypeTok}[1]{\textcolor[rgb]{0.56,0.13,0.00}{{#1}}}
    \newcommand{\DecValTok}[1]{\textcolor[rgb]{0.25,0.63,0.44}{{#1}}}
    \newcommand{\BaseNTok}[1]{\textcolor[rgb]{0.25,0.63,0.44}{{#1}}}
    \newcommand{\FloatTok}[1]{\textcolor[rgb]{0.25,0.63,0.44}{{#1}}}
    \newcommand{\CharTok}[1]{\textcolor[rgb]{0.25,0.44,0.63}{{#1}}}
    \newcommand{\StringTok}[1]{\textcolor[rgb]{0.25,0.44,0.63}{{#1}}}
    \newcommand{\CommentTok}[1]{\textcolor[rgb]{0.38,0.63,0.69}{\textit{{#1}}}}
    \newcommand{\OtherTok}[1]{\textcolor[rgb]{0.00,0.44,0.13}{{#1}}}
    \newcommand{\AlertTok}[1]{\textcolor[rgb]{1.00,0.00,0.00}{\textbf{{#1}}}}
    \newcommand{\FunctionTok}[1]{\textcolor[rgb]{0.02,0.16,0.49}{{#1}}}
    \newcommand{\RegionMarkerTok}[1]{{#1}}
    \newcommand{\ErrorTok}[1]{\textcolor[rgb]{1.00,0.00,0.00}{\textbf{{#1}}}}
    \newcommand{\NormalTok}[1]{{#1}}
    
    % Additional commands for more recent versions of Pandoc
    \newcommand{\ConstantTok}[1]{\textcolor[rgb]{0.53,0.00,0.00}{{#1}}}
    \newcommand{\SpecialCharTok}[1]{\textcolor[rgb]{0.25,0.44,0.63}{{#1}}}
    \newcommand{\VerbatimStringTok}[1]{\textcolor[rgb]{0.25,0.44,0.63}{{#1}}}
    \newcommand{\SpecialStringTok}[1]{\textcolor[rgb]{0.73,0.40,0.53}{{#1}}}
    \newcommand{\ImportTok}[1]{{#1}}
    \newcommand{\DocumentationTok}[1]{\textcolor[rgb]{0.73,0.13,0.13}{\textit{{#1}}}}
    \newcommand{\AnnotationTok}[1]{\textcolor[rgb]{0.38,0.63,0.69}{\textbf{\textit{{#1}}}}}
    \newcommand{\CommentVarTok}[1]{\textcolor[rgb]{0.38,0.63,0.69}{\textbf{\textit{{#1}}}}}
    \newcommand{\VariableTok}[1]{\textcolor[rgb]{0.10,0.09,0.49}{{#1}}}
    \newcommand{\ControlFlowTok}[1]{\textcolor[rgb]{0.00,0.44,0.13}{\textbf{{#1}}}}
    \newcommand{\OperatorTok}[1]{\textcolor[rgb]{0.40,0.40,0.40}{{#1}}}
    \newcommand{\BuiltInTok}[1]{{#1}}
    \newcommand{\ExtensionTok}[1]{{#1}}
    \newcommand{\PreprocessorTok}[1]{\textcolor[rgb]{0.74,0.48,0.00}{{#1}}}
    \newcommand{\AttributeTok}[1]{\textcolor[rgb]{0.49,0.56,0.16}{{#1}}}
    \newcommand{\InformationTok}[1]{\textcolor[rgb]{0.38,0.63,0.69}{\textbf{\textit{{#1}}}}}
    \newcommand{\WarningTok}[1]{\textcolor[rgb]{0.38,0.63,0.69}{\textbf{\textit{{#1}}}}}
    
    
    % Define a nice break command that doesn't care if a line doesn't already
    % exist.
    \def\br{\hspace*{\fill} \\* }
    % Math Jax compatibility definitions
    \def\gt{>}
    \def\lt{<}
    \let\Oldtex\TeX
    \let\Oldlatex\LaTeX
    \renewcommand{\TeX}{\textrm{\Oldtex}}
    \renewcommand{\LaTeX}{\textrm{\Oldlatex}}
    % Document parameters
    % Document title
    \title{2\_fba}
    
    
    
    
    
% Pygments definitions
\makeatletter
\def\PY@reset{\let\PY@it=\relax \let\PY@bf=\relax%
    \let\PY@ul=\relax \let\PY@tc=\relax%
    \let\PY@bc=\relax \let\PY@ff=\relax}
\def\PY@tok#1{\csname PY@tok@#1\endcsname}
\def\PY@toks#1+{\ifx\relax#1\empty\else%
    \PY@tok{#1}\expandafter\PY@toks\fi}
\def\PY@do#1{\PY@bc{\PY@tc{\PY@ul{%
    \PY@it{\PY@bf{\PY@ff{#1}}}}}}}
\def\PY#1#2{\PY@reset\PY@toks#1+\relax+\PY@do{#2}}

\@namedef{PY@tok@w}{\def\PY@tc##1{\textcolor[rgb]{0.73,0.73,0.73}{##1}}}
\@namedef{PY@tok@c}{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.25,0.50,0.50}{##1}}}
\@namedef{PY@tok@cp}{\def\PY@tc##1{\textcolor[rgb]{0.74,0.48,0.00}{##1}}}
\@namedef{PY@tok@k}{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\@namedef{PY@tok@kp}{\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\@namedef{PY@tok@kt}{\def\PY@tc##1{\textcolor[rgb]{0.69,0.00,0.25}{##1}}}
\@namedef{PY@tok@o}{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\@namedef{PY@tok@ow}{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.67,0.13,1.00}{##1}}}
\@namedef{PY@tok@nb}{\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\@namedef{PY@tok@nf}{\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,1.00}{##1}}}
\@namedef{PY@tok@nc}{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,1.00}{##1}}}
\@namedef{PY@tok@nn}{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,1.00}{##1}}}
\@namedef{PY@tok@ne}{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.82,0.25,0.23}{##1}}}
\@namedef{PY@tok@nv}{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\@namedef{PY@tok@no}{\def\PY@tc##1{\textcolor[rgb]{0.53,0.00,0.00}{##1}}}
\@namedef{PY@tok@nl}{\def\PY@tc##1{\textcolor[rgb]{0.63,0.63,0.00}{##1}}}
\@namedef{PY@tok@ni}{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.60,0.60,0.60}{##1}}}
\@namedef{PY@tok@na}{\def\PY@tc##1{\textcolor[rgb]{0.49,0.56,0.16}{##1}}}
\@namedef{PY@tok@nt}{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\@namedef{PY@tok@nd}{\def\PY@tc##1{\textcolor[rgb]{0.67,0.13,1.00}{##1}}}
\@namedef{PY@tok@s}{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\@namedef{PY@tok@sd}{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\@namedef{PY@tok@si}{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.73,0.40,0.53}{##1}}}
\@namedef{PY@tok@se}{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.73,0.40,0.13}{##1}}}
\@namedef{PY@tok@sr}{\def\PY@tc##1{\textcolor[rgb]{0.73,0.40,0.53}{##1}}}
\@namedef{PY@tok@ss}{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\@namedef{PY@tok@sx}{\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\@namedef{PY@tok@m}{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\@namedef{PY@tok@gh}{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,0.50}{##1}}}
\@namedef{PY@tok@gu}{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.50,0.00,0.50}{##1}}}
\@namedef{PY@tok@gd}{\def\PY@tc##1{\textcolor[rgb]{0.63,0.00,0.00}{##1}}}
\@namedef{PY@tok@gi}{\def\PY@tc##1{\textcolor[rgb]{0.00,0.63,0.00}{##1}}}
\@namedef{PY@tok@gr}{\def\PY@tc##1{\textcolor[rgb]{1.00,0.00,0.00}{##1}}}
\@namedef{PY@tok@ge}{\let\PY@it=\textit}
\@namedef{PY@tok@gs}{\let\PY@bf=\textbf}
\@namedef{PY@tok@gp}{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,0.50}{##1}}}
\@namedef{PY@tok@go}{\def\PY@tc##1{\textcolor[rgb]{0.53,0.53,0.53}{##1}}}
\@namedef{PY@tok@gt}{\def\PY@tc##1{\textcolor[rgb]{0.00,0.27,0.87}{##1}}}
\@namedef{PY@tok@err}{\def\PY@bc##1{{\setlength{\fboxsep}{\string -\fboxrule}\fcolorbox[rgb]{1.00,0.00,0.00}{1,1,1}{\strut ##1}}}}
\@namedef{PY@tok@kc}{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\@namedef{PY@tok@kd}{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\@namedef{PY@tok@kn}{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\@namedef{PY@tok@kr}{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\@namedef{PY@tok@bp}{\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\@namedef{PY@tok@fm}{\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,1.00}{##1}}}
\@namedef{PY@tok@vc}{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\@namedef{PY@tok@vg}{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\@namedef{PY@tok@vi}{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\@namedef{PY@tok@vm}{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\@namedef{PY@tok@sa}{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\@namedef{PY@tok@sb}{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\@namedef{PY@tok@sc}{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\@namedef{PY@tok@dl}{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\@namedef{PY@tok@s2}{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\@namedef{PY@tok@sh}{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\@namedef{PY@tok@s1}{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\@namedef{PY@tok@mb}{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\@namedef{PY@tok@mf}{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\@namedef{PY@tok@mh}{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\@namedef{PY@tok@mi}{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\@namedef{PY@tok@il}{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\@namedef{PY@tok@mo}{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\@namedef{PY@tok@ch}{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.25,0.50,0.50}{##1}}}
\@namedef{PY@tok@cm}{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.25,0.50,0.50}{##1}}}
\@namedef{PY@tok@cpf}{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.25,0.50,0.50}{##1}}}
\@namedef{PY@tok@c1}{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.25,0.50,0.50}{##1}}}
\@namedef{PY@tok@cs}{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.25,0.50,0.50}{##1}}}

\def\PYZbs{\char`\\}
\def\PYZus{\char`\_}
\def\PYZob{\char`\{}
\def\PYZcb{\char`\}}
\def\PYZca{\char`\^}
\def\PYZam{\char`\&}
\def\PYZlt{\char`\<}
\def\PYZgt{\char`\>}
\def\PYZsh{\char`\#}
\def\PYZpc{\char`\%}
\def\PYZdl{\char`\$}
\def\PYZhy{\char`\-}
\def\PYZsq{\char`\'}
\def\PYZdq{\char`\"}
\def\PYZti{\char`\~}
% for compatibility with earlier versions
\def\PYZat{@}
\def\PYZlb{[}
\def\PYZrb{]}
\makeatother


    % For linebreaks inside Verbatim environment from package fancyvrb. 
    \makeatletter
        \newbox\Wrappedcontinuationbox 
        \newbox\Wrappedvisiblespacebox 
        \newcommand*\Wrappedvisiblespace {\textcolor{red}{\textvisiblespace}} 
        \newcommand*\Wrappedcontinuationsymbol {\textcolor{red}{\llap{\tiny$\m@th\hookrightarrow$}}} 
        \newcommand*\Wrappedcontinuationindent {3ex } 
        \newcommand*\Wrappedafterbreak {\kern\Wrappedcontinuationindent\copy\Wrappedcontinuationbox} 
        % Take advantage of the already applied Pygments mark-up to insert 
        % potential linebreaks for TeX processing. 
        %        {, <, #, %, $, ' and ": go to next line. 
        %        _, }, ^, &, >, - and ~: stay at end of broken line. 
        % Use of \textquotesingle for straight quote. 
        \newcommand*\Wrappedbreaksatspecials {% 
            \def\PYGZus{\discretionary{\char`\_}{\Wrappedafterbreak}{\char`\_}}% 
            \def\PYGZob{\discretionary{}{\Wrappedafterbreak\char`\{}{\char`\{}}% 
            \def\PYGZcb{\discretionary{\char`\}}{\Wrappedafterbreak}{\char`\}}}% 
            \def\PYGZca{\discretionary{\char`\^}{\Wrappedafterbreak}{\char`\^}}% 
            \def\PYGZam{\discretionary{\char`\&}{\Wrappedafterbreak}{\char`\&}}% 
            \def\PYGZlt{\discretionary{}{\Wrappedafterbreak\char`\<}{\char`\<}}% 
            \def\PYGZgt{\discretionary{\char`\>}{\Wrappedafterbreak}{\char`\>}}% 
            \def\PYGZsh{\discretionary{}{\Wrappedafterbreak\char`\#}{\char`\#}}% 
            \def\PYGZpc{\discretionary{}{\Wrappedafterbreak\char`\%}{\char`\%}}% 
            \def\PYGZdl{\discretionary{}{\Wrappedafterbreak\char`\$}{\char`\$}}% 
            \def\PYGZhy{\discretionary{\char`\-}{\Wrappedafterbreak}{\char`\-}}% 
            \def\PYGZsq{\discretionary{}{\Wrappedafterbreak\textquotesingle}{\textquotesingle}}% 
            \def\PYGZdq{\discretionary{}{\Wrappedafterbreak\char`\"}{\char`\"}}% 
            \def\PYGZti{\discretionary{\char`\~}{\Wrappedafterbreak}{\char`\~}}% 
        } 
        % Some characters . , ; ? ! / are not pygmentized. 
        % This macro makes them "active" and they will insert potential linebreaks 
        \newcommand*\Wrappedbreaksatpunct {% 
            \lccode`\~`\.\lowercase{\def~}{\discretionary{\hbox{\char`\.}}{\Wrappedafterbreak}{\hbox{\char`\.}}}% 
            \lccode`\~`\,\lowercase{\def~}{\discretionary{\hbox{\char`\,}}{\Wrappedafterbreak}{\hbox{\char`\,}}}% 
            \lccode`\~`\;\lowercase{\def~}{\discretionary{\hbox{\char`\;}}{\Wrappedafterbreak}{\hbox{\char`\;}}}% 
            \lccode`\~`\:\lowercase{\def~}{\discretionary{\hbox{\char`\:}}{\Wrappedafterbreak}{\hbox{\char`\:}}}% 
            \lccode`\~`\?\lowercase{\def~}{\discretionary{\hbox{\char`\?}}{\Wrappedafterbreak}{\hbox{\char`\?}}}% 
            \lccode`\~`\!\lowercase{\def~}{\discretionary{\hbox{\char`\!}}{\Wrappedafterbreak}{\hbox{\char`\!}}}% 
            \lccode`\~`\/\lowercase{\def~}{\discretionary{\hbox{\char`\/}}{\Wrappedafterbreak}{\hbox{\char`\/}}}% 
            \catcode`\.\active
            \catcode`\,\active 
            \catcode`\;\active
            \catcode`\:\active
            \catcode`\?\active
            \catcode`\!\active
            \catcode`\/\active 
            \lccode`\~`\~ 	
        }
    \makeatother

    \let\OriginalVerbatim=\Verbatim
    \makeatletter
    \renewcommand{\Verbatim}[1][1]{%
        %\parskip\z@skip
        \sbox\Wrappedcontinuationbox {\Wrappedcontinuationsymbol}%
        \sbox\Wrappedvisiblespacebox {\FV@SetupFont\Wrappedvisiblespace}%
        \def\FancyVerbFormatLine ##1{\hsize\linewidth
            \vtop{\raggedright\hyphenpenalty\z@\exhyphenpenalty\z@
                \doublehyphendemerits\z@\finalhyphendemerits\z@
                \strut ##1\strut}%
        }%
        % If the linebreak is at a space, the latter will be displayed as visible
        % space at end of first line, and a continuation symbol starts next line.
        % Stretch/shrink are however usually zero for typewriter font.
        \def\FV@Space {%
            \nobreak\hskip\z@ plus\fontdimen3\font minus\fontdimen4\font
            \discretionary{\copy\Wrappedvisiblespacebox}{\Wrappedafterbreak}
            {\kern\fontdimen2\font}%
        }%
        
        % Allow breaks at special characters using \PYG... macros.
        \Wrappedbreaksatspecials
        % Breaks at punctuation characters . , ; ? ! and / need catcode=\active 	
        \OriginalVerbatim[#1,codes*=\Wrappedbreaksatpunct]%
    }
    \makeatother

    % Exact colors from NB
    \definecolor{incolor}{HTML}{303F9F}
    \definecolor{outcolor}{HTML}{D84315}
    \definecolor{cellborder}{HTML}{CFCFCF}
    \definecolor{cellbackground}{HTML}{F7F7F7}
    
    % prompt
    \makeatletter
    \newcommand{\boxspacing}{\kern\kvtcb@left@rule\kern\kvtcb@boxsep}
    \makeatother
    \newcommand{\prompt}[4]{
        {\ttfamily\llap{{\color{#2}[#3]:\hspace{3pt}#4}}\vspace{-\baselineskip}}
    }
    

    
    % Prevent overflowing lines due to hard-to-break entities
    \sloppy 
    % Setup hyperref package
    \hypersetup{
      breaklinks=true,  % so long urls are correctly broken across lines
      colorlinks=true,
      urlcolor=urlcolor,
      linkcolor=linkcolor,
      citecolor=citecolor,
      }
    % Slightly bigger margins than the latex defaults
    
    \geometry{verbose,tmargin=1in,bmargin=1in,lmargin=1in,rmargin=1in}
    
    

\begin{document}
    
    \maketitle
    
    

    
    \hypertarget{playing-around-with-media-conditions}{%
\section{2. Playing around with media
conditions}\label{playing-around-with-media-conditions}}

    \hypertarget{authors}{%
\subsection{Authors}\label{authors}}

\begin{itemize}
\tightlist
\item
  Arianna Basile, MRC Toxicology Unit, University of Cambridge
\item
  Francisco Zorrilla, MRC Toxicology Unit, University of Cambridge
\end{itemize}

    \hypertarget{learning-outcomes}{%
\subsection{Learning outcomes}\label{learning-outcomes}}

In this tutorial you will use
\href{https://cobrapy.readthedocs.io/en/latest/}{cobrapy} to learn how
to:

\begin{itemize}
\tightlist
\item
  \textbf{2.1}: Modify growth medium of your reconstruction
\item
  \textbf{2.2}: Perform gene essentiality analysis under different
  conditions
\item
  \textbf{2.3}: Case study, simulate the Crabtree effect in yeast
\end{itemize}

\hypertarget{setup}{%
\subsection{Setup}\label{setup}}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{1}{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\PY{c+c1}{\PYZsh{} Import packages}
\PY{k+kn}{import} \PY{n+nn}{cobra}

\PY{c+c1}{\PYZsh{} Enable autocompleting with tab}
\PY{o}{\PYZpc{}}\PY{k}{config} Completer.use\PYZus{}jedi = False
\end{Verbatim}
\end{tcolorbox}

    \hypertarget{growth-medium}{%
\subsection{2.1 Growth medium}\label{growth-medium}}

    The availability of nutrients has a major impact on metabolic fluxes and
\texttt{cobrapy} provides some helpers to manage the exchanges between
the external environment and your metabolic model. In experimental
settings the ``environment'' is usually constituted by the growth
medium, i.e.~the concentrations of all metabolites and co-factors
available to the modeled organism. However, constraint-based metabolic
models only consider fluxes. Thus, you cannot simply use concentrations
since fluxes have the unit
\texttt{mmol\ /\ {[}gram\ of\ dry-cell\ weight\ *\ hour{]}}
(i.e.~concentration per gram dry weight of cells and hour).

Also, you are setting an upper bound for the particular import flux and
not the flux itself. There are some crude approximations. For instance,
if you supply 1 mmol of glucose every 24h to 1 gram of bacteria you
might set the upper exchange flux for glucose to
\texttt{1\ mmol\ /\ {[}1\ gDW\ *\ h{]}} since that is the nominal
maximum that can be imported. There is no guarantee however that glucose
will be consumed with that flux. Thus, the preferred data for exchange
fluxes are direct flux measurements as the ones obtained from timecourse
exa-metabolome measurements for instance.

So how does that look in COBRApy? The current growth medium of a model
is managed by the medium attribute.

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{2}{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\PY{c+c1}{\PYZsh{} Import model}
\PY{n}{model\PYZus{}yeast}\PY{o}{=}\PY{n}{cobra}\PY{o}{.}\PY{n}{io}\PY{o}{.}\PY{n}{read\PYZus{}sbml\PYZus{}model}\PY{p}{(}\PY{l+s+s2}{\PYZdq{}}\PY{l+s+s2}{iMM904.xml.gz}\PY{l+s+s2}{\PYZdq{}}\PY{p}{)}

\PY{c+c1}{\PYZsh{} Extract media information}
\PY{n}{medium}\PY{o}{=}\PY{n}{model\PYZus{}yeast}\PY{o}{.}\PY{n}{medium}

\PY{c+c1}{\PYZsh{} Show uptake bounds for extracellular media components}
\PY{n+nb}{print}\PY{p}{(}\PY{l+s+s2}{\PYZdq{}}\PY{l+s+s2}{Default extracellular media components for yeast model: }\PY{l+s+s2}{\PYZdq{}}\PY{p}{)}
\PY{n}{medium}
\end{Verbatim}
\end{tcolorbox}

    \begin{Verbatim}[commandchars=\\\{\}]
Default extracellular media components for yeast model:
    \end{Verbatim}

            \begin{tcolorbox}[breakable, size=fbox, boxrule=.5pt, pad at break*=1mm, opacityfill=0]
\prompt{Out}{outcolor}{2}{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\{'EX\_fe2\_e': 999999.0,
 'EX\_glc\_\_D\_e': 10.0,
 'EX\_h2o\_e': 999999.0,
 'EX\_h\_e': 999999.0,
 'EX\_k\_e': 999999.0,
 'EX\_na1\_e': 999999.0,
 'EX\_so4\_e': 999999.0,
 'EX\_nh4\_e': 999999.0,
 'EX\_o2\_e': 2.0,
 'EX\_pi\_e': 999999.0\}
\end{Verbatim}
\end{tcolorbox}
        
    This will return a dictionary that contains the upper flux bounds for
all active exchange fluxes (the ones having non-zero flux bounds). Right
now we see that we have enabled aerobic growth. Let's optimize to check
the growth in the given medium.

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{3}{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\PY{c+c1}{\PYZsh{} Solve FBA}
\PY{n}{model\PYZus{}yeast}\PY{o}{.}\PY{n}{optimize}\PY{p}{(}\PY{p}{)}

\PY{c+c1}{\PYZsh{} Show growth rate with glucose as carbon source}
\PY{n+nb}{print}\PY{p}{(}\PY{l+s+s2}{\PYZdq{}}\PY{l+s+s2}{Growth rate in default medium: }\PY{l+s+s2}{\PYZdq{}}\PY{p}{,}\PY{n}{model\PYZus{}yeast}\PY{o}{.}\PY{n}{objective}\PY{o}{.}\PY{n}{value}\PY{p}{,}\PY{l+s+s2}{\PYZdq{}}\PY{l+s+s2}{mmol/(gDW * h)}\PY{l+s+s2}{\PYZdq{}}\PY{p}{)}
\end{Verbatim}
\end{tcolorbox}

    \begin{Verbatim}[commandchars=\\\{\}]
Growth rate in default medium:  0.28786570370401793 mmol/(gDW * h)
    \end{Verbatim}

    You can modify a growth medium of a model by assigning a dictionary to
\texttt{model.medium} that maps exchange reactions to their respective
upper import bounds. Let's edit the medium to have ethanol as carbon
source instead of glucose.

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{4}{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\PY{c+c1}{\PYZsh{} Make a copy of the original model}
\PY{n}{model\PYZus{}yeastv2}\PY{o}{=}\PY{n}{model\PYZus{}yeast}\PY{o}{.}\PY{n}{copy}\PY{p}{(}\PY{p}{)}

\PY{c+c1}{\PYZsh{} Extract media information}
\PY{n}{medium} \PY{o}{=} \PY{n}{model\PYZus{}yeastv2}\PY{o}{.}\PY{n}{medium}

\PY{c+c1}{\PYZsh{} Define new media uptake bounds}
\PY{n}{medium}\PY{p}{[}\PY{l+s+s2}{\PYZdq{}}\PY{l+s+s2}{EX\PYZus{}glc\PYZus{}\PYZus{}D\PYZus{}e}\PY{l+s+s2}{\PYZdq{}}\PY{p}{]} \PY{o}{=} \PY{l+m+mi}{0}
\PY{n}{medium}\PY{p}{[}\PY{l+s+s2}{\PYZdq{}}\PY{l+s+s2}{EX\PYZus{}etoh\PYZus{}e}\PY{l+s+s2}{\PYZdq{}}\PY{p}{]} \PY{o}{=} \PY{l+m+mf}{1.0}

\PY{c+c1}{\PYZsh{} Update new media uptake bounds in model copy}
\PY{n}{model\PYZus{}yeastv2}\PY{o}{.}\PY{n}{medium} \PY{o}{=} \PY{n}{medium}

\PY{c+c1}{\PYZsh{} Show new uptake bounds}
\PY{n+nb}{print}\PY{p}{(}\PY{l+s+s2}{\PYZdq{}}\PY{l+s+s2}{Modified extracellular media components for yeast model: }\PY{l+s+s2}{\PYZdq{}}\PY{p}{)}
\PY{n}{model\PYZus{}yeastv2}\PY{o}{.}\PY{n}{medium}
\end{Verbatim}
\end{tcolorbox}

    \begin{Verbatim}[commandchars=\\\{\}]
Modified extracellular media components for yeast model:
    \end{Verbatim}

            \begin{tcolorbox}[breakable, size=fbox, boxrule=.5pt, pad at break*=1mm, opacityfill=0]
\prompt{Out}{outcolor}{4}{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\{'EX\_etoh\_e': 1.0,
 'EX\_fe2\_e': 999999.0,
 'EX\_h2o\_e': 999999.0,
 'EX\_h\_e': 999999.0,
 'EX\_k\_e': 999999.0,
 'EX\_na1\_e': 999999.0,
 'EX\_so4\_e': 999999.0,
 'EX\_nh4\_e': 999999.0,
 'EX\_o2\_e': 2.0,
 'EX\_pi\_e': 999999.0\}
\end{Verbatim}
\end{tcolorbox}
        
    As we can see, the glucose in our medium has been replaced by ethanol.
Let's check how it influences the growth rate.

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{5}{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\PY{c+c1}{\PYZsh{} Solve FBA with new uptake bounds}
\PY{n}{model\PYZus{}yeastv2}\PY{o}{.}\PY{n}{optimize}\PY{p}{(}\PY{p}{)}

\PY{c+c1}{\PYZsh{} Show growth rate with ethanol as carbon source}
\PY{n+nb}{print}\PY{p}{(}\PY{l+s+s2}{\PYZdq{}}\PY{l+s+s2}{Growth rate in modified medium with ethanol as carbon source: }\PY{l+s+s2}{\PYZdq{}}\PY{p}{,}
      \PY{n}{model\PYZus{}yeastv2}\PY{o}{.}\PY{n}{objective}\PY{o}{.}\PY{n}{value}\PY{p}{,}\PY{l+s+s2}{\PYZdq{}}\PY{l+s+s2}{mmol/(gDW * h)}\PY{l+s+s2}{\PYZdq{}}\PY{p}{)}
\end{Verbatim}
\end{tcolorbox}

    \begin{Verbatim}[commandchars=\\\{\}]
Growth rate in modified medium with ethanol as carbon source:
0.027542115993883887 mmol/(gDW * h)
    \end{Verbatim}

    The growth rate is reduced by around one order of magnitude when growing
on ethanol compared to glucose. In conclusion, we have the same
metabolic reconstruction (\texttt{iMM904}), simulated under two
different media conditions using different carbon sources.
\texttt{model\_yeast} is growing with glucose as a carbon source while
\texttt{model\_yeastv2} is using ethanol.

\hypertarget{questions}{%
\subsubsection{Questions}\label{questions}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Can the yeast model grow using lactose as a carbon source? What about
  galactose? Was this an expected result?
\item
  What happens when you lower the uptake rate of ammonium? How about
  when it is set to 0? Was this expected?
\item
  Can you quantify the changes in growth rate induced by replacing
  ammonium with an amino acid of your choice to the media?
  e.g.~\texttt{EX\_val\_\_L\_e}
\end{enumerate}

\hypertarget{tip-1}{%
\paragraph{Tip \#1}\label{tip-1}}

\begin{itemize}
\tightlist
\item
  Use the \texttt{model\_yeast.exchanges} command to see which exchange
  reactions are present in the model.
\end{itemize}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{6}{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\PY{c+c1}{\PYZsh{} Show first N exchange reactions}
\PY{n}{N}\PY{o}{=}\PY{l+m+mi}{10}
\PY{n}{model\PYZus{}yeast}\PY{o}{.}\PY{n}{exchanges}\PY{p}{[}\PY{p}{:}\PY{n}{N}\PY{p}{]}
\end{Verbatim}
\end{tcolorbox}

            \begin{tcolorbox}[breakable, size=fbox, boxrule=.5pt, pad at break*=1mm, opacityfill=0]
\prompt{Out}{outcolor}{6}{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
[<Reaction EX\_epistest\_SC\_e at 0x7fead05d9d90>,
 <Reaction EX\_epist\_e at 0x7fead05d9c10>,
 <Reaction EX\_ergst\_e at 0x7fead093ea00>,
 <Reaction EX\_ergstest\_SC\_e at 0x7fead0911e20>,
 <Reaction EX\_13BDglcn\_e at 0x7fead093e640>,
 <Reaction EX\_etha\_e at 0x7fead093eeb0>,
 <Reaction EX\_2hb\_e at 0x7fead093e9a0>,
 <Reaction EX\_etoh\_e at 0x7fead093eca0>,
 <Reaction EX\_fe2\_e at 0x7fead093ef70>,
 <Reaction EX\_fecost\_e at 0x7fead093edc0>]
\end{Verbatim}
\end{tcolorbox}
        
    \hypertarget{tip-2}{%
\paragraph{Tip \#2}\label{tip-2}}

\begin{itemize}
\tightlist
\item
  Use the \href{http://bigg.ucsd.edu/}{BiGG database} to translate
  metabolite/reaction IDs. For example, click
  \href{http://bigg.ucsd.edu/search?query=EX_etha_e}{here} to learn
  about the exchange reaction \texttt{EX\_etha\_e}.
\end{itemize}

\hypertarget{answers}{%
\subsubsection{Answers}\label{answers}}

\hypertarget{question-1}{%
\paragraph{Question 1}\label{question-1}}

Yeast cannot grow on lactose, as it cannot break it down into simple
sugars
{[}\href{https://www.sciencedirect.com/science/article/pii/S2405471217303903}{ref}{]}.
The model reflects this fact, as it is missing an exchange reaction for
lactose. When we try to modify the uptake rate of lactose we get an
error, because the model does not have an exchange reaction for lactose.
On the other hand, yeast can indeed grow on galactose.

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{7}{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\PY{c+c1}{\PYZsh{} Make a copy of the original model}
\PY{n}{model\PYZus{}yeastv3}\PY{o}{=}\PY{n}{model\PYZus{}yeast}\PY{o}{.}\PY{n}{copy}\PY{p}{(}\PY{p}{)}

\PY{c+c1}{\PYZsh{} Extract media information}
\PY{n}{medium} \PY{o}{=} \PY{n}{model\PYZus{}yeastv3}\PY{o}{.}\PY{n}{medium}

\PY{c+c1}{\PYZsh{} Define new media uptake bounds}
\PY{n}{medium}\PY{p}{[}\PY{l+s+s2}{\PYZdq{}}\PY{l+s+s2}{EX\PYZus{}glc\PYZus{}\PYZus{}D\PYZus{}e}\PY{l+s+s2}{\PYZdq{}}\PY{p}{]} \PY{o}{=} \PY{l+m+mi}{0}
\PY{n}{medium}\PY{p}{[}\PY{l+s+s2}{\PYZdq{}}\PY{l+s+s2}{EX\PYZus{}etoh\PYZus{}e}\PY{l+s+s2}{\PYZdq{}}\PY{p}{]} \PY{o}{=} \PY{l+m+mi}{0}
\PY{c+c1}{\PYZsh{}medium[\PYZdq{}EX\PYZus{}lcts\PYZus{}e\PYZdq{}] = 1}
\PY{n}{medium}\PY{p}{[}\PY{l+s+s2}{\PYZdq{}}\PY{l+s+s2}{EX\PYZus{}gal\PYZus{}e}\PY{l+s+s2}{\PYZdq{}}\PY{p}{]} \PY{o}{=} \PY{l+m+mi}{10}

\PY{c+c1}{\PYZsh{} Update new media uptake bounds in model copy}
\PY{n}{model\PYZus{}yeastv3}\PY{o}{.}\PY{n}{medium} \PY{o}{=} \PY{n}{medium}

\PY{c+c1}{\PYZsh{} Show new uptake bounds}
\PY{n+nb}{print}\PY{p}{(}\PY{l+s+s2}{\PYZdq{}}\PY{l+s+s2}{Modified extracellular media components for yeast model with galactose as carbon source: }\PY{l+s+s2}{\PYZdq{}}\PY{p}{)}
\PY{n}{model\PYZus{}yeastv3}\PY{o}{.}\PY{n}{medium}
\end{Verbatim}
\end{tcolorbox}

    \begin{Verbatim}[commandchars=\\\{\}]
Modified extracellular media components for yeast model with galactose as carbon
source:
    \end{Verbatim}

            \begin{tcolorbox}[breakable, size=fbox, boxrule=.5pt, pad at break*=1mm, opacityfill=0]
\prompt{Out}{outcolor}{7}{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\{'EX\_fe2\_e': 999999.0,
 'EX\_gal\_e': 10,
 'EX\_h2o\_e': 999999.0,
 'EX\_h\_e': 999999.0,
 'EX\_k\_e': 999999.0,
 'EX\_na1\_e': 999999.0,
 'EX\_so4\_e': 999999.0,
 'EX\_nh4\_e': 999999.0,
 'EX\_o2\_e': 2.0,
 'EX\_pi\_e': 999999.0\}
\end{Verbatim}
\end{tcolorbox}
        
    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{8}{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\PY{c+c1}{\PYZsh{} Solve FBA with new uptake bounds}
\PY{n}{model\PYZus{}yeastv3}\PY{o}{.}\PY{n}{optimize}\PY{p}{(}\PY{p}{)}

\PY{c+c1}{\PYZsh{} Show growth rate with galactose as carbon source}
\PY{n+nb}{print}\PY{p}{(}\PY{l+s+s2}{\PYZdq{}}\PY{l+s+se}{\PYZbs{}n}\PY{l+s+s2}{Growth rate in modified medium with galactose as carbon source:}\PY{l+s+s2}{\PYZdq{}}\PY{p}{,}
      \PY{n}{model\PYZus{}yeastv3}\PY{o}{.}\PY{n}{objective}\PY{o}{.}\PY{n}{value}\PY{p}{,}\PY{l+s+s2}{\PYZdq{}}\PY{l+s+s2}{mmol/(gDW * h)}\PY{l+s+s2}{\PYZdq{}}\PY{p}{)}
\end{Verbatim}
\end{tcolorbox}

    \begin{Verbatim}[commandchars=\\\{\}]

Growth rate in modified medium with galactose as carbon source:
0.2878657037040199 mmol/(gDW * h)
    \end{Verbatim}

    \hypertarget{question-2}{%
\paragraph{Question 2}\label{question-2}}

Lowering the ammonium uptake rate causes a decrease in growth rate.
Setting the uptake rate to 0 results in near-zero growth rate. This is
expected, as yeast needs a nitrogen source in order to grow
{[}\href{https://www.sciencedirect.com/science/article/pii/S2405471217303903}{ref}{]}.

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{9}{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\PY{c+c1}{\PYZsh{} Make a copy of the original model}
\PY{n}{model\PYZus{}yeastv3}\PY{o}{=}\PY{n}{model\PYZus{}yeast}\PY{o}{.}\PY{n}{copy}\PY{p}{(}\PY{p}{)}

\PY{c+c1}{\PYZsh{} Extract media information}
\PY{n}{medium} \PY{o}{=} \PY{n}{model\PYZus{}yeastv3}\PY{o}{.}\PY{n}{medium}

\PY{c+c1}{\PYZsh{} Define new media uptake bounds}
\PY{n}{medium}\PY{p}{[}\PY{l+s+s2}{\PYZdq{}}\PY{l+s+s2}{EX\PYZus{}glc\PYZus{}\PYZus{}D\PYZus{}e}\PY{l+s+s2}{\PYZdq{}}\PY{p}{]} \PY{o}{=} \PY{l+m+mi}{10}
\PY{n}{medium}\PY{p}{[}\PY{l+s+s2}{\PYZdq{}}\PY{l+s+s2}{EX\PYZus{}etoh\PYZus{}e}\PY{l+s+s2}{\PYZdq{}}\PY{p}{]} \PY{o}{=} \PY{l+m+mi}{0}
\PY{n}{medium}\PY{p}{[}\PY{l+s+s2}{\PYZdq{}}\PY{l+s+s2}{EX\PYZus{}gal\PYZus{}e}\PY{l+s+s2}{\PYZdq{}}\PY{p}{]} \PY{o}{=} \PY{l+m+mi}{0}
\PY{n}{medium}\PY{p}{[}\PY{l+s+s2}{\PYZdq{}}\PY{l+s+s2}{EX\PYZus{}nh4\PYZus{}e}\PY{l+s+s2}{\PYZdq{}}\PY{p}{]} \PY{o}{=} \PY{l+m+mi}{1}
\PY{c+c1}{\PYZsh{}medium[\PYZdq{}EX\PYZus{}nh4\PYZus{}e\PYZdq{}] = 0}

\PY{c+c1}{\PYZsh{} Update new media uptake bounds in model copy}
\PY{n}{model\PYZus{}yeastv3}\PY{o}{.}\PY{n}{medium} \PY{o}{=} \PY{n}{medium}

\PY{c+c1}{\PYZsh{} Show new uptake bounds}
\PY{n+nb}{print}\PY{p}{(}\PY{l+s+s2}{\PYZdq{}}\PY{l+s+s2}{Modified extracellular media components for yeast model with less ammonium: }\PY{l+s+s2}{\PYZdq{}}\PY{p}{)}
\PY{n}{model\PYZus{}yeastv3}\PY{o}{.}\PY{n}{medium}
\end{Verbatim}
\end{tcolorbox}

    \begin{Verbatim}[commandchars=\\\{\}]
Modified extracellular media components for yeast model with less ammonium:
    \end{Verbatim}

            \begin{tcolorbox}[breakable, size=fbox, boxrule=.5pt, pad at break*=1mm, opacityfill=0]
\prompt{Out}{outcolor}{9}{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\{'EX\_fe2\_e': 999999.0,
 'EX\_glc\_\_D\_e': 10,
 'EX\_h2o\_e': 999999.0,
 'EX\_h\_e': 999999.0,
 'EX\_k\_e': 999999.0,
 'EX\_na1\_e': 999999.0,
 'EX\_so4\_e': 999999.0,
 'EX\_nh4\_e': 1,
 'EX\_o2\_e': 2.0,
 'EX\_pi\_e': 999999.0\}
\end{Verbatim}
\end{tcolorbox}
        
    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{10}{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\PY{c+c1}{\PYZsh{} Solve FBA with new uptake bounds}
\PY{n}{model\PYZus{}yeastv3}\PY{o}{.}\PY{n}{optimize}\PY{p}{(}\PY{p}{)}

\PY{c+c1}{\PYZsh{} Show growth rate with less nitrogen}
\PY{n+nb}{print}\PY{p}{(}\PY{l+s+s2}{\PYZdq{}}\PY{l+s+s2}{Growth rate in modified medium with less ammonium:}\PY{l+s+s2}{\PYZdq{}}\PY{p}{,}
      \PY{n}{model\PYZus{}yeastv3}\PY{o}{.}\PY{n}{objective}\PY{o}{.}\PY{n}{value}\PY{p}{,}\PY{l+s+s2}{\PYZdq{}}\PY{l+s+s2}{mmol/(gDW * h)}\PY{l+s+s2}{\PYZdq{}}\PY{p}{)}
\end{Verbatim}
\end{tcolorbox}

    \begin{Verbatim}[commandchars=\\\{\}]
Growth rate in modified medium with less ammonium: 0.17868438262760275 mmol/(gDW
* h)
    \end{Verbatim}

    \hypertarget{question-3}{%
\paragraph{Question 3}\label{question-3}}

Replacing ammonium with valine results in \textasciitilde2.2\% increase
in growth rate.

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{11}{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\PY{c+c1}{\PYZsh{} Make a copy of the original model}
\PY{n}{model\PYZus{}yeastv3}\PY{o}{=}\PY{n}{model\PYZus{}yeast}\PY{o}{.}\PY{n}{copy}\PY{p}{(}\PY{p}{)}

\PY{c+c1}{\PYZsh{} Extract media information}
\PY{n}{medium} \PY{o}{=} \PY{n}{model\PYZus{}yeastv3}\PY{o}{.}\PY{n}{medium}

\PY{c+c1}{\PYZsh{} Define new media uptake bounds}
\PY{n}{medium}\PY{p}{[}\PY{l+s+s2}{\PYZdq{}}\PY{l+s+s2}{EX\PYZus{}glc\PYZus{}\PYZus{}D\PYZus{}e}\PY{l+s+s2}{\PYZdq{}}\PY{p}{]} \PY{o}{=} \PY{l+m+mi}{10}
\PY{n}{medium}\PY{p}{[}\PY{l+s+s2}{\PYZdq{}}\PY{l+s+s2}{EX\PYZus{}etoh\PYZus{}e}\PY{l+s+s2}{\PYZdq{}}\PY{p}{]} \PY{o}{=} \PY{l+m+mi}{0}
\PY{n}{medium}\PY{p}{[}\PY{l+s+s2}{\PYZdq{}}\PY{l+s+s2}{EX\PYZus{}gal\PYZus{}e}\PY{l+s+s2}{\PYZdq{}}\PY{p}{]} \PY{o}{=} \PY{l+m+mi}{0}
\PY{n}{medium}\PY{p}{[}\PY{l+s+s2}{\PYZdq{}}\PY{l+s+s2}{EX\PYZus{}nh4\PYZus{}e}\PY{l+s+s2}{\PYZdq{}}\PY{p}{]} \PY{o}{=} \PY{l+m+mi}{0}
\PY{n}{medium}\PY{p}{[}\PY{l+s+s2}{\PYZdq{}}\PY{l+s+s2}{EX\PYZus{}val\PYZus{}\PYZus{}L\PYZus{}e}\PY{l+s+s2}{\PYZdq{}}\PY{p}{]} \PY{o}{=} \PY{l+m+mi}{1000}

\PY{c+c1}{\PYZsh{} Update new media uptake bounds in model copy}
\PY{n}{model\PYZus{}yeastv3}\PY{o}{.}\PY{n}{medium} \PY{o}{=} \PY{n}{medium}

\PY{c+c1}{\PYZsh{} Show new uptake bounds}
\PY{n+nb}{print}\PY{p}{(}\PY{l+s+s2}{\PYZdq{}}\PY{l+s+s2}{Modified extracellular media components for yeast model with valine as nitrogen source: }\PY{l+s+s2}{\PYZdq{}}\PY{p}{)}
\PY{n}{model\PYZus{}yeastv3}\PY{o}{.}\PY{n}{medium}
\end{Verbatim}
\end{tcolorbox}

    \begin{Verbatim}[commandchars=\\\{\}]
Modified extracellular media components for yeast model with valine as nitrogen
source:
    \end{Verbatim}

            \begin{tcolorbox}[breakable, size=fbox, boxrule=.5pt, pad at break*=1mm, opacityfill=0]
\prompt{Out}{outcolor}{11}{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\{'EX\_fe2\_e': 999999.0,
 'EX\_glc\_\_D\_e': 10,
 'EX\_h2o\_e': 999999.0,
 'EX\_h\_e': 999999.0,
 'EX\_k\_e': 999999.0,
 'EX\_na1\_e': 999999.0,
 'EX\_so4\_e': 999999.0,
 'EX\_o2\_e': 2.0,
 'EX\_pi\_e': 999999.0,
 'EX\_val\_\_L\_e': 1000\}
\end{Verbatim}
\end{tcolorbox}
        
    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{12}{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\PY{c+c1}{\PYZsh{} Solve FBA with new uptake bounds}
\PY{n}{model\PYZus{}yeastv3}\PY{o}{.}\PY{n}{optimize}\PY{p}{(}\PY{p}{)}

\PY{c+c1}{\PYZsh{} Show growth rate with valine as nitrogen source}
\PY{n+nb}{print}\PY{p}{(}\PY{l+s+s2}{\PYZdq{}}\PY{l+s+s2}{Growth rate in modified medium with valine as nitrogen source:}\PY{l+s+s2}{\PYZdq{}}\PY{p}{,}
      \PY{n}{model\PYZus{}yeastv3}\PY{o}{.}\PY{n}{objective}\PY{o}{.}\PY{n}{value}\PY{p}{,}\PY{l+s+s2}{\PYZdq{}}\PY{l+s+s2}{mmol/(gDW * h)}\PY{l+s+s2}{\PYZdq{}}\PY{p}{)}

\PY{c+c1}{\PYZsh{} Calculate as percentage}
\PY{n+nb}{print}\PY{p}{(}\PY{l+s+s2}{\PYZdq{}}\PY{l+s+s2}{Growth rate on valine as nitrogen source relative to ammonium:}\PY{l+s+s2}{\PYZdq{}}\PY{p}{,}
      \PY{p}{(}\PY{n}{model\PYZus{}yeastv3}\PY{o}{.}\PY{n}{objective}\PY{o}{.}\PY{n}{value}\PY{o}{/}\PY{n}{model\PYZus{}yeast}\PY{o}{.}\PY{n}{objective}\PY{o}{.}\PY{n}{value}\PY{p}{)}\PY{o}{*}\PY{l+m+mi}{100}\PY{p}{,}\PY{l+s+s2}{\PYZdq{}}\PY{l+s+s2}{\PYZpc{}}\PY{l+s+s2}{\PYZdq{}}\PY{p}{)}
\end{Verbatim}
\end{tcolorbox}

    \begin{Verbatim}[commandchars=\\\{\}]
Growth rate in modified medium with valine as nitrogen source:
0.2942390011149862 mmol/(gDW * h)
Growth rate on valine as nitrogen source relative to ammonium:
102.21398288471393 \%
    \end{Verbatim}

    \hypertarget{gene-essentiality}{%
\subsection{2.2 Gene essentiality}\label{gene-essentiality}}

    A gene is considered essential if restricting the flux of all reactions
that depend on it to zero causes the objective, e.g., the growth rate,
to also be zero, fall below a given threshold, or result in infeasible
solutions.

Now we will use the function \texttt{find\_essential\_genes} to find
essential genes in both conditions, extract the IDs of the genes, and
compare the two lists. Can you do it on your own?

First lets check the model growing with ethanol as a carbon source:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{13}{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\PY{c+c1}{\PYZsh{} Find essential genes for model growing under ethanol condition}
\PY{n}{essential\PYZus{}genes\PYZus{}v2}\PY{o}{=}\PY{n}{cobra}\PY{o}{.}\PY{n}{flux\PYZus{}analysis}\PY{o}{.}\PY{n}{variability}\PY{o}{.}\PY{n}{find\PYZus{}essential\PYZus{}genes}\PY{p}{(}\PY{n}{model\PYZus{}yeastv2}\PY{p}{)}

\PY{c+c1}{\PYZsh{} Get genes IDs}
\PY{n}{id\PYZus{}essential\PYZus{}genes\PYZus{}v2}\PY{o}{=}\PY{p}{[}\PY{n}{gene}\PY{o}{.}\PY{n}{id} \PY{k}{for} \PY{n}{gene} \PY{o+ow}{in} \PY{n}{essential\PYZus{}genes\PYZus{}v2}\PY{p}{]}

\PY{c+c1}{\PYZsh{} Print number of esssential genes}
\PY{n+nb}{print}\PY{p}{(}\PY{l+s+s2}{\PYZdq{}}\PY{l+s+s2}{Essential genes with ethanol carbon source:}\PY{l+s+s2}{\PYZdq{}}\PY{p}{,}\PY{n+nb}{len}\PY{p}{(}\PY{n}{id\PYZus{}essential\PYZus{}genes\PYZus{}v2}\PY{p}{)}\PY{p}{)}
\end{Verbatim}
\end{tcolorbox}

    \begin{Verbatim}[commandchars=\\\{\}]
Essential genes with ethanol carbon source: 154
    \end{Verbatim}

    Next lets check the model growing with glucose as a carbon source:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{14}{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\PY{c+c1}{\PYZsh{} Find essential genes for model growing under glucose condition}
\PY{n}{essential\PYZus{}genes}\PY{o}{=}\PY{n}{cobra}\PY{o}{.}\PY{n}{flux\PYZus{}analysis}\PY{o}{.}\PY{n}{variability}\PY{o}{.}\PY{n}{find\PYZus{}essential\PYZus{}genes}\PY{p}{(}\PY{n}{model\PYZus{}yeast}\PY{p}{)}

\PY{c+c1}{\PYZsh{} Get genes IDs}
\PY{n}{id\PYZus{}essential\PYZus{}genes}\PY{o}{=}\PY{p}{[}\PY{n}{gene}\PY{o}{.}\PY{n}{id} \PY{k}{for} \PY{n}{gene} \PY{o+ow}{in} \PY{n}{essential\PYZus{}genes}\PY{p}{]}

\PY{c+c1}{\PYZsh{} Print number of essential genes}
\PY{n+nb}{print}\PY{p}{(}\PY{l+s+s2}{\PYZdq{}}\PY{l+s+s2}{Essential genes with glucose carbon source:}\PY{l+s+s2}{\PYZdq{}}\PY{p}{,}\PY{n+nb}{len}\PY{p}{(}\PY{n}{id\PYZus{}essential\PYZus{}genes}\PY{p}{)}\PY{p}{)}
\end{Verbatim}
\end{tcolorbox}

    \begin{Verbatim}[commandchars=\\\{\}]
Essential genes with glucose carbon source: 110
    \end{Verbatim}

    Check how large the intersect is between the two lists:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{15}{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\PY{n}{a}\PY{o}{=}\PY{n+nb}{list}\PY{p}{(}\PY{n+nb}{set}\PY{p}{(}\PY{n}{id\PYZus{}essential\PYZus{}genes\PYZus{}v2}\PY{p}{)} \PY{o}{\PYZam{}} \PY{n+nb}{set}\PY{p}{(}\PY{n}{id\PYZus{}essential\PYZus{}genes}\PY{p}{)}\PY{p}{)}
\PY{n+nb}{print}\PY{p}{(}\PY{l+s+s2}{\PYZdq{}}\PY{l+s+s2}{Intersect of essential genes between conditions:}\PY{l+s+s2}{\PYZdq{}}\PY{p}{,}\PY{n+nb}{len}\PY{p}{(}\PY{n}{a}\PY{p}{)}\PY{p}{)}
\end{Verbatim}
\end{tcolorbox}

    \begin{Verbatim}[commandchars=\\\{\}]
Intersect of essential genes between conditions: 110
    \end{Verbatim}

    Lets check which genes are unique to each condition:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{16}{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\PY{n+nb}{print}\PY{p}{(}\PY{l+s+s2}{\PYZdq{}}\PY{l+s+s2}{Essential genes unique to ethanol growth condition:}\PY{l+s+s2}{\PYZdq{}}\PY{p}{,}
      \PY{n+nb}{len}\PY{p}{(}\PY{n+nb}{list}\PY{p}{(}\PY{n+nb}{set}\PY{p}{(}\PY{n}{id\PYZus{}essential\PYZus{}genes\PYZus{}v2}\PY{p}{)} \PY{o}{\PYZhy{}} \PY{n+nb}{set}\PY{p}{(}\PY{n}{id\PYZus{}essential\PYZus{}genes}\PY{p}{)}\PY{p}{)}\PY{p}{)}\PY{p}{)}
\PY{n+nb}{list}\PY{p}{(}\PY{n+nb}{set}\PY{p}{(}\PY{n}{id\PYZus{}essential\PYZus{}genes\PYZus{}v2}\PY{p}{)} \PY{o}{\PYZhy{}} \PY{n+nb}{set}\PY{p}{(}\PY{n}{id\PYZus{}essential\PYZus{}genes}\PY{p}{)}\PY{p}{)}
\end{Verbatim}
\end{tcolorbox}

    \begin{Verbatim}[commandchars=\\\{\}]
Essential genes unique to ethanol growth condition: 44
    \end{Verbatim}

            \begin{tcolorbox}[breakable, size=fbox, boxrule=.5pt, pad at break*=1mm, opacityfill=0]
\prompt{Out}{outcolor}{16}{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
['YKL060C',
 'YEL024W',
 'YDL181W',
 'YHR051W',
 'YBR039W',
 'YBR196C',
 'YLR295C',
 'YLR395C',
 'YLL041C',
 'YJR121W',
 'YLR038C',
 'YBL099W',
 'YGR183C',
 'YDR298C',
 'Q0130',
 'YPR191W',
 'YKL016C',
 'YER065C',
 'YKR097W',
 'YPL262W',
 'YDR529C',
 'YDL004W',
 'YDL067C',
 'YDR377W',
 'Q0250',
 'YGL187C',
 'YDR050C',
 'Q0275',
 'YBL045C',
 'YJL166W',
 'YCR012W',
 'Q0085',
 'Q0080',
 'YML081C\_A',
 'YPL271W',
 'YLR377C',
 'YPL078C',
 'YMR256C',
 'YGL191W',
 'Q0045',
 'YHR001W\_A',
 'YOR065W',
 'YFR033C',
 'Q0105']
\end{Verbatim}
\end{tcolorbox}
        
    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{17}{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\PY{n+nb}{print}\PY{p}{(}\PY{l+s+s2}{\PYZdq{}}\PY{l+s+s2}{Essential genes unique to glucose growth condition:}\PY{l+s+s2}{\PYZdq{}}\PY{p}{,}
      \PY{n+nb}{len}\PY{p}{(}\PY{n+nb}{list}\PY{p}{(}\PY{n+nb}{set}\PY{p}{(}\PY{n}{id\PYZus{}essential\PYZus{}genes}\PY{p}{)} \PY{o}{\PYZhy{}} \PY{n+nb}{set}\PY{p}{(}\PY{n}{id\PYZus{}essential\PYZus{}genes\PYZus{}v2}\PY{p}{)}\PY{p}{)}\PY{p}{)}\PY{p}{)}
\PY{n+nb}{list}\PY{p}{(}\PY{n+nb}{set}\PY{p}{(}\PY{n}{id\PYZus{}essential\PYZus{}genes}\PY{p}{)} \PY{o}{\PYZhy{}} \PY{n+nb}{set}\PY{p}{(}\PY{n}{id\PYZus{}essential\PYZus{}genes\PYZus{}v2}\PY{p}{)}\PY{p}{)}
\end{Verbatim}
\end{tcolorbox}

    \begin{Verbatim}[commandchars=\\\{\}]
Essential genes unique to glucose growth condition: 0
    \end{Verbatim}

            \begin{tcolorbox}[breakable, size=fbox, boxrule=.5pt, pad at break*=1mm, opacityfill=0]
\prompt{Out}{outcolor}{17}{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
[]
\end{Verbatim}
\end{tcolorbox}
        
    \hypertarget{questions}{%
\subsubsection{Questions}\label{questions}}

    \begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Why there are more essential genes for growth on ethanol than for
  growth on glucose?
\item
  What pathways are these ethanol-growth essential genes associated
  with?
\item
  Choose an essential gene unique to growth in ethanol and search the
  literature for its function. Does it make sense to you that it is
  essential?
\end{enumerate}

\hypertarget{tip}{%
\paragraph{Tip}\label{tip}}

For an overview of central carbon metabolism in yeast refer to the
figure below, obtained from
\href{https://journals.asm.org/doi/10.1128/mbio.02970-21}{this
publication}.

\begin{figure}
\centering
\includegraphics{attachment:image.png}
\caption{image.png}
\end{figure}

\hypertarget{answers}{%
\subsubsection{Answers}\label{answers}}

\hypertarget{question-1}{%
\paragraph{Question 1}\label{question-1}}

Because growing on ethanol requires an extra set of enzymes which are
not needed for growth on glucose. Growing on glucose requires enzymes
for glycolysis only, while growing on ethanol additionally requires
aerobic fermentation enzymes.

\hypertarget{question-2}{%
\paragraph{Question 2}\label{question-2}}

These genes are associated with the TCA cycle and oxidative
phosphorylation (i.e.~electron transport chain).

\hypertarget{question-3}{%
\paragraph{Question 3}\label{question-3}}

A search for the gene
\href{https://www.yeastgenome.org/locus/S000004387}{YLR395C} reveals
that it is indeed associated with the electron transport chain and
cellular respiration.

    \hypertarget{simulating-the-crabtree-effect}{%
\subsection{2.3 Simulating the Crabtree
effect}\label{simulating-the-crabtree-effect}}

    Yeast central carbon metabolic pathways: all short-term Crabtree
positive yeasts possess an upregulated aerobic (blue) and anaerobic
(red) glycolytic pathway, even under fully aerobic conditions, when
energy and carbon-source is limiting. At low glucose uptake rates,
yeasts cells are purely respiring and there is a carbon-flux only
through glycolysis (GF) and respiration (RF). Upon a sudden glucose
excess condition (glucose pulse), the glycolytic flux will exceed the
respiratory flux, which results in a fermentative flux (FF) and ethanol
production. In other words, it appears as if slowly dividing short-term
Crabtree positive cells, with little food around are already equipped
with a strong energy producing apparatus, for rapid glucose consumption
and energy production.

\includegraphics{attachment:image-3.png} (Adapted from Hagman and
Piškur, 2015)

    To simulate the Crabtree effect, we consider a fixed amount of glucose
and different concentrations of oxygen in our medium.

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{18}{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\PY{c+c1}{\PYZsh{} Create copy of model}
\PY{n}{model\PYZus{}yeastv3}\PY{o}{=}\PY{n}{model\PYZus{}yeast}\PY{o}{.}\PY{n}{copy}\PY{p}{(}\PY{p}{)}

\PY{c+c1}{\PYZsh{} Define range of oxygen uptake rates}
\PY{n}{oxygen\PYZus{}uptakes}\PY{o}{=}\PY{n+nb}{range}\PY{p}{(}\PY{l+m+mi}{1}\PY{p}{,}\PY{l+m+mi}{1000}\PY{p}{,}\PY{l+m+mi}{10}\PY{p}{)}
\PY{n}{oxygen\PYZus{}uptakes} \PY{o}{=} \PY{p}{[} \PY{n}{x} \PY{o}{/} \PY{l+m+mi}{100} \PY{k}{for} \PY{n}{x} \PY{o+ow}{in} \PY{n}{oxygen\PYZus{}uptakes}\PY{p}{]}

\PY{c+c1}{\PYZsh{} Define ethanol production vector}
\PY{n}{etoh\PYZus{}prod}\PY{o}{=}\PY{p}{[}\PY{p}{]}

\PY{c+c1}{\PYZsh{} Simulate under varying oxygen uptake rates, fixed glucose}
\PY{k}{for} \PY{n}{flux} \PY{o+ow}{in} \PY{n}{oxygen\PYZus{}uptakes}\PY{p}{:}
    \PY{n}{medium} \PY{o}{=} \PY{n}{model\PYZus{}yeastv3}\PY{o}{.}\PY{n}{medium}
    \PY{n}{medium}\PY{p}{[}\PY{l+s+s2}{\PYZdq{}}\PY{l+s+s2}{EX\PYZus{}glc\PYZus{}\PYZus{}D\PYZus{}e}\PY{l+s+s2}{\PYZdq{}}\PY{p}{]} \PY{o}{=} \PY{l+m+mi}{15}
    \PY{n}{medium}\PY{p}{[}\PY{l+s+s2}{\PYZdq{}}\PY{l+s+s2}{EX\PYZus{}o2\PYZus{}e}\PY{l+s+s2}{\PYZdq{}}\PY{p}{]}\PY{o}{=}\PY{n}{flux}
    \PY{n}{model\PYZus{}yeastv3}\PY{o}{.}\PY{n}{medium} \PY{o}{=} \PY{n}{medium}
    \PY{n}{opt}\PY{o}{=}\PY{n}{model\PYZus{}yeastv3}\PY{o}{.}\PY{n}{optimize}\PY{p}{(}\PY{p}{)}
    \PY{n}{etoh\PYZus{}prod}\PY{o}{.}\PY{n}{append}\PY{p}{(}\PY{n}{opt}\PY{o}{.}\PY{n}{fluxes}\PY{p}{[}\PY{l+s+s2}{\PYZdq{}}\PY{l+s+s2}{EX\PYZus{}etoh\PYZus{}e}\PY{l+s+s2}{\PYZdq{}}\PY{p}{]}\PY{p}{)}
\end{Verbatim}
\end{tcolorbox}

    Plot results

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{19}{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\PY{c+c1}{\PYZsh{} Import plotting library}
\PY{k+kn}{import} \PY{n+nn}{matplotlib}\PY{n+nn}{.}\PY{n+nn}{pyplot} \PY{k}{as} \PY{n+nn}{plt}

\PY{c+c1}{\PYZsh{} Define plot}
\PY{n}{x\PYZus{}axis}\PY{o}{=}\PY{n}{oxygen\PYZus{}uptakes}
\PY{n}{y\PYZus{}axis}\PY{o}{=}\PY{n}{etoh\PYZus{}prod}
\PY{n}{plt}\PY{o}{.}\PY{n}{plot}\PY{p}{(}\PY{n}{x\PYZus{}axis}\PY{p}{,} \PY{n}{y\PYZus{}axis}\PY{p}{)}
\PY{n}{plt}\PY{o}{.}\PY{n}{title}\PY{p}{(}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{Crabtree effect in yeast model}\PY{l+s+s1}{\PYZsq{}}\PY{p}{)}
\PY{n}{plt}\PY{o}{.}\PY{n}{xlabel}\PY{p}{(}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{Oxygen uptake mmol / [gDW * h]}\PY{l+s+s1}{\PYZsq{}}\PY{p}{)}
\PY{n}{plt}\PY{o}{.}\PY{n}{ylabel}\PY{p}{(}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{Ethanol production}\PY{l+s+s1}{\PYZsq{}}\PY{p}{)}
\PY{n}{plt}\PY{o}{.}\PY{n}{show}\PY{p}{(}\PY{p}{)}
\end{Verbatim}
\end{tcolorbox}

    \begin{center}
    \adjustimage{max size={0.9\linewidth}{0.9\paperheight}}{output_41_0.png}
    \end{center}
    { \hspace*{\fill} \\}
    
    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{20}{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\PY{c+c1}{\PYZsh{} Type your code here if you want to try similar analyses using different carbon sources, e.g. D\PYZhy{}mannose}
\end{Verbatim}
\end{tcolorbox}

    \hypertarget{questions}{%
\subsubsection{Questions}\label{questions}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Can you explain the observed behavior in the plot above?
\item
  Try playing around with other media components, for example, what
  happens if you keep oxygen fixed and vary glucose?
\item
  Literature search question: What are two similarities and two
  differences between the Crabtree and Warburg effects?
\end{enumerate}

\hypertarget{tip}{%
\paragraph{Tip}\label{tip}}

Check out
\href{https://www.tandfonline.com/doi/full/10.1080/15384101.2018.1442622}{this
publication} describing the Crabtree and Warburg effects. In particular,
have a look at Figure 1 shown below:

\begin{figure}
\centering
\includegraphics{attachment:image-2.png}
\caption{image-2.png}
\end{figure}

\hypertarget{answers}{%
\subsubsection{Answers}\label{answers}}

\hypertarget{question-1}{%
\paragraph{Question 1}\label{question-1}}

The plot shows the Crabtree effect. When glucose is present in excess
and under anaerobic/low oxygen conditions, we see that ethanol
production reaches a maximum. Furthermore, we see that as oxygen uptake
increases, yeast reduces its fermentative flux but it does not reach
zero.

\hypertarget{question-2}{%
\paragraph{Question 2}\label{question-2}}

One of the conditions of the Crabtree effect is that the carbon source
(i.e.~glucose) has to be present in excess. In line with this
precondition, we see that very little ethanol is produced (i.e.~low
fermentative flux) until glucose is no longer limiting.

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{21}{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\PY{c+c1}{\PYZsh{} Create copy of model}
\PY{n}{model\PYZus{}yeastv3}\PY{o}{=}\PY{n}{model\PYZus{}yeast}\PY{o}{.}\PY{n}{copy}\PY{p}{(}\PY{p}{)}

\PY{c+c1}{\PYZsh{} Define range of oxygen uptake rates}
\PY{n}{glucose\PYZus{}uptakes}\PY{o}{=}\PY{n+nb}{range}\PY{p}{(}\PY{l+m+mi}{1}\PY{p}{,}\PY{l+m+mi}{1000}\PY{p}{,}\PY{l+m+mi}{10}\PY{p}{)}
\PY{n}{glucose\PYZus{}uptakes} \PY{o}{=} \PY{p}{[} \PY{n}{x} \PY{o}{/} \PY{l+m+mi}{100} \PY{k}{for} \PY{n}{x} \PY{o+ow}{in} \PY{n}{glucose\PYZus{}uptakes}\PY{p}{]}

\PY{c+c1}{\PYZsh{} Define ethanol production vector}
\PY{n}{etoh\PYZus{}prod}\PY{o}{=}\PY{p}{[}\PY{p}{]}

\PY{c+c1}{\PYZsh{} Simulate under varying oxygen uptake rates, fixed glucose}
\PY{k}{for} \PY{n}{flux} \PY{o+ow}{in} \PY{n}{glucose\PYZus{}uptakes}\PY{p}{:}
    \PY{n}{medium} \PY{o}{=} \PY{n}{model\PYZus{}yeastv3}\PY{o}{.}\PY{n}{medium}
    \PY{n}{medium}\PY{p}{[}\PY{l+s+s2}{\PYZdq{}}\PY{l+s+s2}{EX\PYZus{}o2\PYZus{}e}\PY{l+s+s2}{\PYZdq{}}\PY{p}{]} \PY{o}{=} \PY{l+m+mi}{10}
    \PY{n}{medium}\PY{p}{[}\PY{l+s+s2}{\PYZdq{}}\PY{l+s+s2}{EX\PYZus{}glc\PYZus{}\PYZus{}D\PYZus{}e}\PY{l+s+s2}{\PYZdq{}}\PY{p}{]}\PY{o}{=}\PY{n}{flux}
    \PY{n}{model\PYZus{}yeastv3}\PY{o}{.}\PY{n}{medium} \PY{o}{=} \PY{n}{medium}
    \PY{n}{opt}\PY{o}{=}\PY{n}{model\PYZus{}yeastv3}\PY{o}{.}\PY{n}{optimize}\PY{p}{(}\PY{p}{)}
    \PY{n}{etoh\PYZus{}prod}\PY{o}{.}\PY{n}{append}\PY{p}{(}\PY{n}{opt}\PY{o}{.}\PY{n}{fluxes}\PY{p}{[}\PY{l+s+s2}{\PYZdq{}}\PY{l+s+s2}{EX\PYZus{}etoh\PYZus{}e}\PY{l+s+s2}{\PYZdq{}}\PY{p}{]}\PY{p}{)}
    
\PY{c+c1}{\PYZsh{} Define plot}
\PY{n}{x\PYZus{}axis}\PY{o}{=}\PY{n}{oxygen\PYZus{}uptakes}
\PY{n}{y\PYZus{}axis}\PY{o}{=}\PY{n}{etoh\PYZus{}prod}
\PY{n}{plt}\PY{o}{.}\PY{n}{plot}\PY{p}{(}\PY{n}{x\PYZus{}axis}\PY{p}{,} \PY{n}{y\PYZus{}axis}\PY{p}{)}
\PY{n}{plt}\PY{o}{.}\PY{n}{title}\PY{p}{(}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{Crabtree effect in yeast model}\PY{l+s+s1}{\PYZsq{}}\PY{p}{)}
\PY{n}{plt}\PY{o}{.}\PY{n}{xlabel}\PY{p}{(}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{Glucose uptake mmol / [gDW * h]}\PY{l+s+s1}{\PYZsq{}}\PY{p}{)}
\PY{n}{plt}\PY{o}{.}\PY{n}{ylabel}\PY{p}{(}\PY{l+s+s1}{\PYZsq{}}\PY{l+s+s1}{Ethanol production}\PY{l+s+s1}{\PYZsq{}}\PY{p}{)}
\PY{n}{plt}\PY{o}{.}\PY{n}{show}\PY{p}{(}\PY{p}{)}
\end{Verbatim}
\end{tcolorbox}

    \begin{Verbatim}[commandchars=\\\{\}]
/Users/fzorrilla/.local/lib/python3.8/site-packages/cobra/util/solver.py:554:
UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '\{status\}'.", UserWarning)
    \end{Verbatim}

    \begin{center}
    \adjustimage{max size={0.9\linewidth}{0.9\paperheight}}{output_44_1.png}
    \end{center}
    { \hspace*{\fill} \\}
    
    \hypertarget{question-3}{%
\paragraph{Question 3}\label{question-3}}

Similarities: * Both result in high fermentative flux even under aerobic
conditions * Both result in lower ATP yield

Differences: * The Crabtree effect is observed in yeast cells, while the
Warburg effect occurs in mammalian cells * The Crabtree effect produces
ethanol, while the Warburg effect produces lactate


    % Add a bibliography block to the postdoc
    
    
    
\end{document}
